src/Compilers/CSharp/Portable/Binder/Binder_Conversions.cs

```diff
@@ -1022,250 +1015,222 @@ private ImmutableArray<BoundNode> ConvertCollectionExpressionElements(
-            {
-                var elementConversions = conversion.UnderlyingConversions;
 
-                Debug.Assert(elementType is { });
-                Debug.Assert(elementConversions.All(c => c.Exists));
-                Debug.Assert(implicitReceiver is { });
+            Debug.Assert(conversionIndex == elementConversions.Length);
+            conversion.MarkUnderlyingConversionsChecked();
 
-                int conversionIndex = 0;
-                foreach (var element in elements)
+            return builder.ToImmutableAndFree();
+
+            static BoundNode convertKeyValuePair(
+                Binder binder,
+                BoundKeyValuePairElement keyValuePairElement,
+                Conversion elementConversion,
+                TypeSymbol elementType,
+                (TypeSymbol Key, TypeSymbol Value)? elementKeyValueTypes,
+                bool useIndexer,
+                BoundObjectOrCollectionValuePlaceholder? implicitReceiver,
+                CollectionInitializerAddMethodBinder? collectionInitializerAddMethodBinder,
+                BindingDiagnosticBag diagnostics)
+            {
+                if (elementKeyValueTypes is (var elementKeyType, var elementValueType) &&
+                    elementConversion.TryGetKeyValueConversions(out var keyConversion, out var valueConversion))
                 {
-                    if (element is BoundCollectionExpressionWithElement)
+                    var keyValuePairSyntax = (KeyValuePairElementSyntax)keyValuePairElement.Syntax;
+                    var keyValuePairConstructor = useIndexer ? null : (MethodSymbol?)binder.GetWellKnownTypeMember(WellKnownMember.System_Collections_Generic_KeyValuePair_KV__ctor, diagnostics, syntax: keyValuePairSyntax);
+                    var key = binder.CreateConversion(keyValuePairElement.Key, keyConversion, elementKeyType, diagnostics);
+                    var value = binder.CreateConversion(keyValuePairElement.Value, valueConversion, elementValueType, diagnostics);
+                    if (collectionInitializerAddMethodBinder is { })
                     {
-                        continue;
+#if DEBUG
+                        Debug.Assert(ConversionsBase.IsKeyValuePairType(binder.Compilation, elementType, out var keyType, out var valueType) &&
+                            keyType.Equals(key.Type, TypeCompareKind.AllIgnoreOptions) &&
+                            valueType.Equals(value.Type, TypeCompareKind.AllIgnoreOptions));
+#endif
+                        BoundExpression keyValuePair = (keyValuePairConstructor is null) ?
+                            new BoundBadExpression(
+                                keyValuePairSyntax,
+                                LookupResultKind.Empty,
+                                symbols: [],
+                                childBoundNodes: [],
```

Comment:
Should we store the bound parts (`key` and `value`) inside this `BoundBadExpression`?